pkg/util/keymutex/keymutex.go

```diff
@@ -18,16 +18,47 @@ package keymutex
 
 import (
 	"fmt"
-	"sync"
-
 	"github.com/golang/glog"
+	"sync"
+	"sync/atomic"
 )
 
+type tryMutex struct {
+	sync.Mutex
+	acq int32
+}
+
+func (tm *tryMutex) Lock() {
+	atomic.StoreInt32(&tm.acq, 1)
+	tm.Mutex.Lock()
+}
+
+func (tm *tryMutex) TryLock() bool {
```

Comment:
Is there a case where `TryLock` would block? Two callers, one calls `Lock`, the other `TryLock`. If the compare-and-swap returns `true`, now you have a race between two calls to `tm.Mutex.Lock()`, which the `TryLock` call could lose and thus get blocked.

Reply:
That is absolutely true! This was really wrongly done, my bad.